code/__HELPERS/matrices.dm

```diff
@@ -63,3 +63,92 @@
 //The Y pixel offset of this matrix
 /matrix/proc/get_y_shift()
 	. = f
+
+/datum/ColorMatrix
+	var/list/matrix
+	var/combined = 1
+	var/const/lumR = 0.3086 //  or  0.2125
+	var/const/lumG = 0.6094 //  or  0.7154
+	var/const/lumB = 0.0820 //  or  0.0721
+
+/datum/ColorMatrix/New(mat, contrast = 1, brightness = null)
+	..()
+	if(istext(mat))
+		SetColor(mat, contrast, brightness)
+	else if(isnum(mat))
+		SetSaturation(mat, contrast, brightness)
+	else
+		matrix = mat
+
+/datum/ColorMatrix/proc/Reset()
+	matrix=list(1,0,0,
+				0,1,0,
+				0,0,1)
+
+/datum/ColorMatrix/proc/Get(contrast = 1)
+	var/list/mat = matrix
+	mat = mat.Copy()
+
+	for(var/i = 1 to min(mat.len, 12))
+		mat[i] *= contrast
+	return mat
+
+/datum/ColorMatrix/proc/SetSaturation(s, c = 1, b = null)
+	var/sr = (1 - s) * lumR
+	var/sg = (1 - s) * lumG
+	var/sb = (1 - s) * lumB
+
+	matrix=list(c * (sr + s), c * (sr),     c * (sr),
+				c * (sg),     c * (sg + s), c * (sg),
+				c * (sb),     c * (sb),     c * (sb + s))
+	SetBrightness(b)
+
+/datum/ColorMatrix/proc/SetBrightness(brightness)
+	if(brightness == null) return
+
+	if(!matrix)
+		Reset()
+
+	if(matrix.len == 9 || matrix.len == 16)
+		matrix += brightness
+		matrix += brightness
+		matrix += brightness
+
+		if(matrix.len == 16)
+			matrix += 0
+
+	else if(matrix.len == 12)
+		for(var/i = matrix.len to matrix.len - 3 step -1)
+			matrix[i] = brightness
+
+	else if(matrix.len == 3)
+		for(var/i = matrix.len - 1 to matrix.len - 4 step -1)
+			matrix[i] = brightness
+
+/datum/ColorMatrix/proc/hex2value(hex)
+	var/num1 = copytext(hex, 1, 2)
+	var/num2 = copytext(hex, 2)
+	if(isnum(text2num(num1)))
+		num1 = text2num(num1)
+	else
+		num1 = text2ascii(lowertext(num1)) - 87
+	if(isnum(text2num(num1)))
+		num2 = text2num(num1)
+	else
+		num2 = text2ascii(lowertext(num2)) - 87
```

Comment:
```suggestion
		num1 = text2ascii(lowertext(num1)) - text2ascii("a") + 1
	if(isnum(text2num(num1)))
		num2 = text2num(num1)
	else
		num2 = text2ascii(lowertext(num2)) - text2ascii("a") + 1
```

Magic numbers are bad. Pre-calculate and save `text2ascii("a") - 1` if you have performance concerns, but I doubt it's significant.